File: js/digits.js
'use strict'

const TOP = 'top'
const TOP_LEFT = 'topLeft'
const TOP_RIGHT = 'topRight'
const MIDDLE = 'middle'
const BOTTOM_LEFT = 'bottomLeft'
const BOTTOM_RIGHT = 'bottomRight'
const BOTTOM = 'bottom'

const SIDES = [
  TOP, TOP_LEFT, TOP_RIGHT, MIDDLE, BOTTOM_LEFT, BOTTOM_RIGHT, BOTTOM
]

const DIGIT_INACTIVE_SIDES = {
  '0': [MIDDLE],
  '1': [TOP, TOP_LEFT, MIDDLE, BOTTOM_LEFT, BOTTOM],
  '2': [TOP_LEFT, BOTTOM_RIGHT],
  '3': [TOP_LEFT, BOTTOM_LEFT],
  '4': [TOP, BOTTOM_LEFT, BOTTOM],
  '5': [TOP_RIGHT, BOTTOM_LEFT],
  '6': [TOP_RIGHT],
  '7': [MIDDLE, BOTTOM_LEFT, BOTTOM],
  '8': [],
  '9': [BOTTOM_LEFT]
}

const makeDigitObject = sidesForDigit => {
  const digit = {}
  for (const side of SIDES) digit[side] = sidesForDigit.includes(side)
  return digit
}

const digits = {
  '-': makeDigitObject([MIDDLE])
}

for (const [digit, inactiveSides] of Object.entries(DIGIT_INACTIVE_SIDES)) {
  const activeSides = SIDES.filter(side => !inactiveSides.includes(side))
  digits[digit] = makeDigitObject(activeSides)
}

module.exports = digits

File: py/seven_segment_numbers/digits.py
class _Digit:
    TOP = 'top'
    TOP_LEFT = 'top_left'
    TOP_RIGHT = 'top_right'
    MIDDLE = 'middle'
    BOTTOM_LEFT = 'bottom_left'
    BOTTOM_RIGHT = 'bottom_right'
    BOTTOM = 'bottom'

    SIDES = [
        TOP, TOP_LEFT, TOP_RIGHT, MIDDLE, BOTTOM_LEFT, BOTTOM_RIGHT, BOTTOM
    ]

    def __init__(self, sides):
        for side in _Digit.SIDES:
            setattr(self, side, side in sides)

_DIGIT_INACTIVES_SIDES = {
    '0': [_Digit.MIDDLE],
    '1': [
        _Digit.TOP,
        _Digit.TOP_LEFT,
        _Digit.MIDDLE,
        _Digit.BOTTOM_LEFT,
        _Digit.BOTTOM
    ],
    '2': [_Digit.TOP_LEFT, _Digit.BOTTOM_RIGHT],
    '3': [_Digit.TOP_LEFT, _Digit.BOTTOM_LEFT],
    '4': [_Digit.TOP, _Digit.BOTTOM_LEFT, _Digit.BOTTOM],
    '5': [_Digit.TOP_RIGHT, _Digit.BOTTOM_LEFT],
    '6': [_Digit.TOP_RIGHT],
    '7': [_Digit.MIDDLE, _Digit.BOTTOM_LEFT, _Digit.BOTTOM],
    '8': [],
    '9': [_Digit.BOTTOM_LEFT]
}

DIGITS = {
    '-': _Digit([_Digit.MIDDLE])
}

for digit, inactive_sides in _DIGIT_INACTIVES_SIDES.items():
    sides = [side for side in _Digit.SIDES if side not in inactive_sides]
    DIGITS[digit] = _Digit(sides)
